test/unit/serverless/aws-lambda-stream-response.test.js

```diff
@@ -0,0 +1,1855 @@
+  const plan = tspl(t, { plan: 8 })
+  const { agent, awsLambda, error, event, responseStream, context } = t.nr
+
+  agent.on('harvestStarted', function confirmErrorCapture() {
+    const errors = agent.errors.traceAggregator.errors
+    plan.equal(errors.length, 1)
+
+    const noticedError = errors[0]
+    const [, transactionName, message, type] = noticedError
+    plan.equal(transactionName, expectedBgTransactionName)
+    plan.equal(message, errorMessage)
+    plan.equal(type, 'SyntaxError')
+  })
+
+  const handler = decorateHandler(async (event, responseStream) => {
+    const transaction = agent.tracer.getTransaction()
+    const chunks = ['step 1', 'step 2', 'step 3']
+    await writeToResponseStream(chunks, responseStream, 100)
+    responseStream.end()
+    plan.ok(transaction)
+    plan.equal(transaction.type, 'bg')
+    plan.equal(transaction.getFullName(), expectedBgTransactionName)
+    plan.ok(transaction.isActive())
+
+    throw error
+  })
+
+  const wrappedHandler = awsLambda.patchLambdaHandler(handler)
+
+  try {
+    await wrappedHandler(event, responseStream, context)
+  } catch (error) {
+    if (error.name !== 'SyntaxError') {
+      throw error
+    }
+  }
+  await plan.completed
+})
+
+test('should not end transactions twice', async (t) => {
+  const plan = tspl(t, { plan: 6 })
+  const { agent, awsLambda, event, responseStream, context } = t.nr
+  let transaction
+
+  const handler = decorateHandler(async (event, responseStream) => {
+    transaction = agent.tracer.getTransaction()
+    const chunks = ['step 1', 'step 2', 'step 3']
+    await writeToResponseStream(chunks, responseStream, 100)
+    responseStream.end()
+
+    let called = false
+    const oldEnd = transaction.end
+    transaction.end = function wrappedEnd() {
+      if (called) {
+        throw new Error('called end on the same transaction twice')
+      }
+      called = true
+      return oldEnd.apply(transaction, arguments)
+    }
+
+    return new Promise((resolve) => {
+      plan.ok(transaction)
+      plan.equal(transaction.type, 'bg')
+      plan.equal(transaction.getFullName(), expectedBgTransactionName)
+      plan.ok(transaction.isActive())
+      responseStream.end()
+      resolve('hello')
+    })
+  })
+
+  const wrappedHandler = awsLambda.patchLambdaHandler(handler)
+
+  const value = await wrappedHandler(event, responseStream, context)
+  plan.equal(value, 'hello')
+  plan.equal(transaction.isActive(), false)
+  await plan.completed
+})
+
+test('should record standard background metrics', async (t) => {
+  const plan = tspl(t, { plan: 9 })
+  const { agent, awsLambda, event, responseStream, context } = t.nr
+  agent.on('harvestStarted', confirmMetrics)
+
+  const handler = decorateHandler(async (event, responseStream) => {
+    const chunks = ['step 1', 'step 2', 'step 3']
+    await writeToResponseStream(chunks, responseStream, 100)
+    responseStream.end()
+    return validResponse
```

Comment:
I'm not sure this matters but we're not checking that the response stream returns the right data.  You could add `responseDone` to the test context and after you end the stream 

````
const data = await responseDone
plan.equal(data, 'step1step2step3')
```

Reply:
That makes sense to check against what the final stream value is, though with this usage I'm seeing responseDone as undefined.